### DISCUSSION TITLE
Not able to get Force value using sponge.GetForce()

### DISCUSSION THREAD

**Discussion by rajas1310** [empriselab/RCareWorld#169] [Help - Programming] [answered]
Hi,
My code follows the template given for `test_bathing_force.py` to get the force values.
I use :
```
force = sponge.GetForce()
env.step()
print(force)
```

But the printed output shows [0.0] even when the Unity Simulation Window shows a positive "Average Force".
Due to this issue, I am not able to use the force value as a feedback for my robot. How can I solve this issue?

Thank you.

**Comment by YoruCathy** [comment]
Hi, are you on the latest commit of `phy-robo-care`?

**Reply by rajas1310** [reply]
Hi,
Yes I am on commit `72af472`.

**Reply by YoruCathy** [reply]
Does it work on the commit `7cb63a1`?

**Reply by rajas1310** [reply]
No, it isn't working on that commit either.

**Reply by thisjustin123** [reply]
Hello! Currently, the sponge only registers force with the human body. Are you testing force readings against that?

If not, it'd be useful to hear from you if you'd find it useful for us to add sponge force readings against all objects in the scene?

**Reply by rajas1310** [reply]
Hi,
I am trying to get the readings against the human body itself. I am not sure why I am getting them as 0.0 even though the unity window shows a non-zero "Average force"

**Reply by thisjustin123** [reply]
Hmmm very strange. I just tested it and I can correctly read the forces as mostly not 0.0.

You're getting every single force reading on python, without exception, to be 0.0? Is there ever a reading that is NOT 0.0? I get a 0.0 every now and then, but certainly not on every reading...

^ Follow up, what's the **framerate** (estimate is ok) of your unity exectuable? Is it very laggy?

**Reply by rajas1310** [reply]
Yeah, I am getting every reading as 0.0
How can I check the framerate? The simulation doesn't seem to be laggy.

**Reply by rajas1310** [reply]
> Hmmm very strange. I just tested it and I can correctly read the forces as mostly not 0.0.
> 
> You're getting every single force reading on python, without exception, to be 0.0? Is there ever a reading that is NOT 0.0? I get a 0.0 every now and then, but certainly not on every reading...

This is on the latest commit, right?

**Reply by thisjustin123** [reply]
Hmm, as long as your simulation doesn't seem to be laggy that shouldn't be the issue then... I can't say I'm exactly sure what's causing this?

I am testing on the latest commit, but against our Unity Editor (because my machine isn't linux and can't run the executable 😭). Perhaps then there's something wrong with the executable itself being able to read the forces. So it is working in the internal editor, but maybe just not the exectuable?

Sorry to ask, but can you provide a brief demo video (perhaps just a screen cap) showing the sponge pressing against the person with the force in sim going up, but the python output being still 0? If that's too much, just providing the code snippet you're running with a screenshot is also good. I'll try to look into it further with that, but since I'm not sure why that's happening I am unsure of a timeline of a fix.

**Reply by rajas1310** [reply]
Sure,
PFA the video of my run.
The code snippet for where I read the forces is:
```
while force < 5.5 and y_ht > 0.75:
    y_ht -= 0.03
    robot.IKTargetDoMove([point[0], y_ht, point[1]], duration=1, speed_based=False)
    env.step(10)
    force = sponge.GetForce()[0]
    env.step()
    print("---->>>> In while loop\ty_ht =", round(y_ht, 2), "Force: ", force, "N")
```
https://github.com/user-attachments/assets/b21e677e-f818-4c8d-9373-2eb27f447b42

**Reply by thisjustin123** [reply]
Your code snippet seems okay, and we are testing with the exact same function, so I'm not sure where the discrepancy is...

In the meantime, can you try a few things?

1. Please try running `git lfs install` `git lfs pull` if you haven't already. This caused discrepancies / failures running in the past (although I'd imagine your executable would not even run if this was the case, so this may not do anything)
2. Try using `sponge.data["forces"]` to access a list of forces and see if that gives anything.

If neither of those work, we can try to investigate further and perhaps write some test cases, since there could be some weird discrepancy between the Unity Editor and the Executable that we aren't quite catching atm.

**Reply by rajas1310** [reply]
Sure, I will try the above-mentioned things and let you know if any of them work. Thank you, Justin!

**Reply by rajas1310** [reply]
>1. Please try running git lfs install git lfs pull if you haven't already. This caused discrepancies / failures running in the past (although I'd imagine your executable would not even run if this was the case, so this may not do anything)

`rajas@tea ~/p/R/template (phy-robo-care) [1]> git lfs install `    
```
Updated Git hooks.
Git LFS initialized.
```
`rajas@tea ~/p/R/template (phy-robo-care)> git lfs pull`

```<No output>```

>2. Try using sponge.data["forces"] to access a list of forces and see if that gives anything.
The `sponge.data` dict has no key named `"forces"`

PS: I am on the latest commit `78452cc`

**Reply by thisjustin123** [reply]
I'm currently writing a test (and fixing some force code) on https://github.com/empriselab/RCareWorld/pull/172.

The test case is *now* passing when it wasn't before. I'm not sure if this is because I fixed something or just fixed the test case. Either way, if the force reading works in the test cases, I'd assume it would work the same in your sim. 

Feel free to try to hop on that branch and see if your code works, or wait until the PR is merged to test. 😄

**Reply by YoruCathy** [reply, marked as answer]
#172 is merged! Feel free to try the newest commit
